crypto/engine/eng_devcrypto.c

```diff
@@ -471,14 +661,9 @@ static int digest_copy(EVP_MD_CTX *to, const EVP_MD_CTX *from)
         (struct digest_ctx *)EVP_MD_CTX_md_data(to);
     struct cphash_op cphash;
 
-    if (digest_from == NULL)
+    if (digest_from == NULL || digest_from->init != 1)
         return 1;
 
-    if (digest_from->init != 1) {
-        SYSerr(SYS_F_IOCTL, EINVAL);
-        return 0;
-    }
-
```

Comment:
I'm surprised by this change...  do we really try to copy uninitialized `EVP_MD_CTX`'s?

That has me thinking, though, that the `digest_ctx->init = 1;` assignment in `digest_init` maye have to move to be done just before `return 1;`.  What do you think?

Reply:
Yes, we do try to copy uninitialized EVP_MD_CTX.  I was surprised as well, but that's the one that basically fixed the tests.  As for the `digest_ctx->init`, here's what I think:
At first, I had gotten rid of it, as we never checked its value.
Now, we have a single check at `digest_copy`, where if it is not 1, then we don't initialize the destination ctx on copy.  I'm not sure what the best way to handle a copy of a CTX with an initialization failure, but either we move it to the end as you propose, or we rename it to `init_called`, for example, to make clear it is not signaling a successful init.

Reply:
I think renaming it is the best solution.  I imagine this scenario.  We create an MD_CTX, initialized, even though it failed, we can use a copy of it that was successfully initialized.  In fact, IIRC, the operation: _copy an unused CTX (before the first update), update & final the copy, then just close the original_ happens during TLS handshake. I'd be surprised if the init failure would not generate an error before all that happens.

Reply:
I just rant the whole test suite with the `digest_from->init_called` check removed, and it passes,  so I'm not sure what failures you got or why.  Is there some configuration I need to be aware of to be able to reproduce it?

Reply:
```
Test Summary Report
-------------------
../test/recipes/70-test_clienthello.t            (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
../test/recipes/80-test_ssl_new.t                (Wstat: 1024 Tests: 28 Failed: 4)
  Failed tests:  9-10, 12, 27
  Non-zero exit status: 4
../test/recipes/90-test_sslapi.t                 (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
../test/recipes/90-test_tls13ccs.t               (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
Files=152, Tests=1361, 489 wallclock secs ( 1.51 usr  0.20 sys + 148.99 cusr 317.11 csys = 467.81 CPU)
Result: FAIL
```
Mine fails.  Make sure there's an increase in `csys` as above, as that's the indication that the engine is doing its part.

Reply:
The failure was different then.  I ran `LD_LIBRARY_PATH=. gdb --args test/clienthellotest test/session.pem` and set a breakpoint where the old `digest_from->init !=0` test would be true.

Reply:
Here's the failure with current master:
```
        ok 3 - iteration 3
        # ERROR: (bool) 'PACKET_forward(&pkt, SSL3_HM_HEADER_LENGTH) == true' failed @ test/clienthellotest.c:191
        # false
        # 140527382172672:error:02015016:system library:ioctl:Invalid argument:crypto/engine/eng_devcrypto.c:478:
        # 140527382172672:error:141FA044:SSL routines:tls_psk_do_binder:internal error:ssl/statem/extensions.c:1595:
        not ok 4 - iteration 4
```
`eng_devcrypto.c:478` is inside the `if (digest_from->init != 1)`

Reply:
I'd like to debug this for myself, so if you could please tell me the following:

- the output from `perl configdata.pm -c`
- the exact command(s) you used for testing

Reply:
```

 $ perl configdata.pm -c

Command line (with current working directory = .):

    /usr/bin/perl ./Configure debug-linux-x86_64 enable-devcryptoeng --strict-warnings --openssldir=/home/equeiroz/ssl/

Perl information:

    /usr/bin/perl
    5.24.3 for x86_64-linux
```
- Then the test I'm running is (result is with current master, 65d2c16):
```
l $ make test TESTS=test_clienthello V=1
make depend && make _tests
make[1]: Entering directory '/home/equeiroz/src/new/openssl'
make[1]: Leaving directory '/home/equeiroz/src/new/openssl'
make[1]: Entering directory '/home/equeiroz/src/new/openssl'
( cd test; \
  mkdir -p test-runs; \
  SRCTOP=../. \
  BLDTOP=../. \
  RESULT_D=test-runs \
  PERL="/usr/bin/perl" \
  EXE_EXT= \
  OPENSSL_ENGINES=`cd .././engines 2>/dev/null && pwd` \
  OPENSSL_DEBUG_MEMORY=on \
    /usr/bin/perl .././test/run_tests.pl test_clienthello )
../test/recipes/70-test_clienthello.t ..
1..1
    # Subtest: ../../test/clienthellotest
    1..1
        # Subtest: test_client_hello
        1..4
        ok 1 - iteration 1
        ok 2 - iteration 2
        ok 3 - iteration 3
        # ERROR: (bool) 'PACKET_forward(&pkt, SSL3_HM_HEADER_LENGTH) == true' failed @ test/clienthellotest.c:191
        # false
        # 140526593852416:error:02015016:system library:ioctl:Invalid argument:crypto/engine/eng_devcrypto.c:478:
        # 140526593852416:error:141FA044:SSL routines:tls_psk_do_binder:internal error:ssl/statem/extensions.c:1595:
        not ok 4 - iteration 4
    not ok 1 - test_client_hello
../../util/shlib_wrap.sh ../../test/clienthellotest ../../test/session.pem => 1
not ok 1 - running clienthellotest

#   Failed test 'running clienthellotest'
#   at ../test/recipes/70-test_clienthello.t line 20.
# Looks like you failed 1 test of 1.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests

Test Summary Report
-------------------
../test/recipes/70-test_clienthello.t (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
Files=1, Tests=1,  1 wallclock secs ( 0.02 usr  0.00 sys +  0.06 cusr  0.03 csys =  0.11 CPU)
Result: FAIL
make[1]: *** [Makefile:205: _tests] Error 1
make[1]: Leaving directory '/home/equeiroz/src/new/openssl'
make: *** [Makefile:202: tests] Error 2
```
I've traced the algorithms used in the test, and they are `AES-256-ECB` and `SHA384`, so make sure you have kernel drivers for them.  Here's my DUMP_INFO output:
```
$ LD_LIBRARY_PATH=. apps/openssl engine -t -c -v  -pre DUMP_INFO devcrypto
(devcrypto) /dev/crypto engine
Information about ciphers supported by the /dev/crypto engine:
CIOCGSESSINFO (session info call) unavailable
Cipher DES-CBC, NID=31, /dev/crypto info: id=1, driver=unknown (acceleration status unknown)
Cipher DES-EDE3-CBC, NID=44, /dev/crypto info: id=2, driver=unknown (acceleration status unknown)
Cipher BF-CBC, NID=91, /dev/crypto info: id=3, driver=unknown (acceleration status unknown)
Cipher CAST5-CBC, NID=108, /dev/crypto info: id=4, CIOCGSESSION (session open call) failed
Cipher AES-128-CBC, NID=419, /dev/crypto info: id=11, driver=unknown (acceleration status unknown)
Cipher AES-192-CBC, NID=423, /dev/crypto info: id=11, driver=unknown (acceleration status unknown)
Cipher AES-256-CBC, NID=427, /dev/crypto info: id=11, driver=unknown (acceleration status unknown)
Cipher RC4, NID=5, /dev/crypto info: id=12, CIOCGSESSION (session open call) failed
Cipher AES-128-CTR, NID=904, /dev/crypto info: id=21, driver=unknown (acceleration status unknown)
Cipher AES-192-CTR, NID=905, /dev/crypto info: id=21, driver=unknown (acceleration status unknown)
Cipher AES-256-CTR, NID=906, /dev/crypto info: id=21, driver=unknown (acceleration status unknown)
Cipher AES-128-ECB, NID=418, /dev/crypto info: id=23, driver=unknown (acceleration status unknown)
Cipher AES-192-ECB, NID=422, /dev/crypto info: id=23, driver=unknown (acceleration status unknown)
Cipher AES-256-ECB, NID=426, /dev/crypto info: id=23, driver=unknown (acceleration status unknown)
Cipher CAMELLIA-128-CBC, NID=751, /dev/crypto info: id=101, driver=unknown (acceleration status unknown)
Cipher CAMELLIA-192-CBC, NID=752, /dev/crypto info: id=101, driver=unknown (acceleration status unknown)
Cipher CAMELLIA-256-CBC, NID=753, /dev/crypto info: id=101, driver=unknown (acceleration status unknown)

Information about digests supported by the /dev/crypto engine (with the CIOCGSESSINFO hack):
CIOCGSESSINFO (session info call) unavailable
Digest MD5, NID=4, /dev/crypto info: id=13, driver=unknown (acceleration status unknown), CIOCCPHASH capable
Digest SHA1, NID=64, /dev/crypto info: id=14, driver=unknown (acceleration status unknown), CIOCCPHASH capable
Digest RIPEMD160, NID=117, /dev/crypto info: id=102, driver=unknown (acceleration status unknown), CIOCCPHASH capable
Digest SHA224, NID=675, /dev/crypto info: id=103, driver=unknown (acceleration status unknown), CIOCCPHASH capable
Digest SHA256, NID=672, /dev/crypto info: id=104, driver=unknown (acceleration status unknown), CIOCCPHASH capable
Digest SHA384, NID=673, /dev/crypto info: id=105, driver=unknown (acceleration status unknown), CIOCCPHASH capable
Digest SHA512, NID=674, /dev/crypto info: id=106, driver=unknown (acceleration status unknown), CIOCCPHASH capable

[Success]: DUMP_INFO
 [DES-CBC, DES-EDE3-CBC, BF-CBC, AES-128-CBC, AES-192-CBC, AES-256-CBC, AES-128-CTR, AES-192-CTR, AES-256-CTR, AES-128-ECB, AES-192-ECB, AES-256-ECB, CAMELLIA-128-CBC, CAMELLIA-192-CBC, CAMELLIA-256-CBC, MD5, SHA1, RIPEMD160, SHA224, SHA256, SHA384, SHA512]
     [ available ]
     USE_SOFTDRIVERS, CIPHERS, DIGESTS, DUMP_INFO
```

Reply:
I'll add that if you are in doubt if you're using the engine or not, `make test TESTS=test_ssl_new V=1` is useful, as it takes forever to complete with the engine on, and you can't miss the difference in system time (`csys`).
Cheers

Reply:
:laughing: 

Maybe...  just maaaaaaybe...  I should install an updated cryptodev.h.

(of course I get no error when CIOCCPHASH isn't defined :wink:)

Reply:
Ok, seen it for myself, thought about it, and I agree that your `init_called` check is the sanest way.